### ISSUE TITLE
adapter::delete causes a program crash

### ISSUE THREAD

**Issue by pradt2** [nulldotblack/wireguard-nt#3]
Hi, when I run the `demo_server.rs` example from current master (Win 10, MSVC compiler, Rust 1.56), the program crashes when `adapter::delete` is called:

Here's the stacktrace:

```
thread 'main' panicked at 'Expected function, got error.: GetProcAddress { source: Os { code: 127, kind: Uncategorized, message: "The specified procedure could not be found." } }', C:\<fakepath>\wireguard-nt\src\wireguard_nt_raw.rs:1459:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'Expected function, got error.: GetProcAddress { source: Os { code: 127, kind: Uncategorized, message: "The specified procedure could not be found." } }', C:\<fakepath>\wireguard-nt\src\wireguard_nt_raw.rs:1476:14
stack backtrace:
   0:     0x7ff7034c09de - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:98
   1:     0x7ff7034c09de - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff7034c09de - std::sys_common::backtrace::_print_fmt
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\sys_common\backtrace.rs:67
   3:     0x7ff7034c09de - std::sys_common::backtrace::_print::impl$0::fmt
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\sys_common\backtrace.rs:46
   4:     0x7ff7034d03ea - core::fmt::write
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\fmt\mod.rs:1150
   5:     0x7ff7034bb8b8 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\io\mod.rs:1667
   6:     0x7ff7034c2c86 - std::sys_common::backtrace::_print
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\sys_common\backtrace.rs:49
   7:     0x7ff7034c2c86 - std::sys_common::backtrace::print
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\sys_common\backtrace.rs:36
   8:     0x7ff7034c2c86 - std::panicking::default_hook::closure$1
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:210
   9:     0x7ff7034c2774 - std::panicking::default_hook
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:227
  10:     0x7ff7034c32e5 - std::panicking::rust_panic_with_hook
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:624
  11:     0x7ff7034c2ecb - std::panicking::begin_panic_handler::closure$0
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:521
  12:     0x7ff7034c1327 - std::sys_common::backtrace::__rust_end_short_backtrace<std::panicking::begin_panic_handler::closure$0,never$>
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\sys_common\backtrace.rs:141
  13:     0x7ff7034c2e29 - std::panicking::begin_panic_handler
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:517
  14:     0x7ff7034d5b30 - core::panicking::panic_fmt
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:101
  15:     0x7ff7034d5bd3 - core::result::unwrap_failed
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\result.rs:1617
  16:     0x7ff7034316f3 - enum$<core::result::Result<ref$<void (*)(ptr_mut$<wireguard_nt::wireguard_nt_raw::_WIREGUARD_ADAPTER>)>,ref$<enum$<libloading::error::Error> > > >::expect<ref$<void (*)(ptr_mut$<wireguard_nt::wireguard_nt_raw::_WIREGUARD_ADAPTER>)>,ref$<enum$<libloading::
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\result.rs:1259
  17:     0x7ff70342d9d1 - wireguard_nt::wireguard_nt_raw::wireguard::WireGuardFreeAdapter
                               at C:\<fakepath>\wireguard-nt\src\wireguard_nt_raw.rs:1473
  18:     0x7ff703425a0e - wireguard_nt::adapter::impl$1::drop
                               at C:\<fakepath>\wireguard-nt\src\adapter.rs:527
  19:     0x7ff7034250af - core::ptr::drop_in_place<wireguard_nt::adapter::Adapter>
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\ptr\mod.rs:188
  20:     0x7ff703427ba7 - wireguard_nt::adapter::Adapter::delete
                               at C:\<fakepath>\wireguard-nt\src\adapter.rs:514
  21:     0x7fff1c911030 - <unknown>
  22:     0x7fff1c914cb2 - is_exception_typeof
  23:     0x7fff1c91f0e4 - _C_specific_handler
  24:     0x7fff1c913f28 - is_exception_typeof
  25:     0x7fff1c91fa21 - _CxxFrameHandler3
  26:     0x7fff2a9f214f - _chkstk
  27:     0x7fff2a980939 - RtlUnwindEx
  28:     0x7fff1c91f5ae - _C_specific_handler
  29:     0x7fff1c912bf5 - is_exception_typeof
  30:     0x7fff1c91300d - is_exception_typeof
  31:     0x7fff1c914024 - is_exception_typeof
  32:     0x7fff1c91fa21 - _CxxFrameHandler3
  33:     0x7fff2a9f20cf - _chkstk
  34:     0x7fff2a9a1454 - RtlRaiseException
  35:     0x7fff2a9a11a5 - RtlRaiseException
  36:     0x7fff28404f99 - RaiseException
  37:     0x7fff1c916480 - CxxThrowException
  38:     0x7ff7034c7281 - panic_unwind::real_imp::panic
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\panic_unwind\src\seh.rs:313
  39:     0x7ff7034c7209 - panic_unwind::__rust_start_panic
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\panic_unwind\src\lib.rs:105
  40:     0x7ff7034c359b - std::panicking::rust_panic
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:672
  41:     0x7ff7034c349b - std::panicking::rust_panic_with_hook
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:642
  42:     0x7ff7034c2ecb - std::panicking::begin_panic_handler::closure$0
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:521
  43:     0x7ff7034c1327 - std::sys_common::backtrace::__rust_end_short_backtrace<std::panicking::begin_panic_handler::closure$0,never$>
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\sys_common\backtrace.rs:141
  44:     0x7ff7034c2e29 - std::panicking::begin_panic_handler
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:517
  45:     0x7ff7034d5b30 - core::panicking::panic_fmt
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:101
  46:     0x7ff7034d5bd3 - core::result::unwrap_failed
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\result.rs:1617
  47:     0x7ff703431bc3 - enum$<core::result::Result<ref$<i32 (*)(ptr_mut$<wireguard_nt::wireguard_nt_raw::_WIREGUARD_ADAPTER>,ptr_mut$<i32>)>,ref$<enum$<libloading::error::Error> > > >::expect<ref$<i32 (*)(ptr_mut$<wireguard_nt::wireguard_nt_raw::_WIREGUARD_ADAPTER>,ptr_mut$<i32>
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\result.rs:1259
  48:     0x7ff70342d958 - wireguard_nt::wireguard_nt_raw::wireguard::WireGuardDeleteAdapter
                               at C:\<fakepath>\wireguard-nt\src\wireguard_nt_raw.rs:1456
  49:     0x7ff703427b51 - wireguard_nt::adapter::Adapter::delete
                               at C:\<fakepath>\wireguard-nt\src\adapter.rs:505
  50:     0x7ff7034229f1 - my_project::wireguard::WireguardIface::create_interface
                               at C:\<fakepath>\my_project\src\wireguard.rs:127
  51:     0x7ff703422ede - my_project::main
                               at C:\<fakepath>\my_project\src\main.rs:64
  52:     0x7ff7034218db - core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\ops\function.rs:227
  53:     0x7ff70342146b - std::sys_common::backtrace::__rust_begin_short_backtrace<void (*)(),tuple$<> >
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\std\src\sys_common\backtrace.rs:125
  54:     0x7ff703421191 - std::rt::lang_start::closure$0<tuple$<> >
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\std\src\rt.rs:63
  55:     0x7ff7034c3736 - core::ops::function::impls::impl$2::call_once
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\ops\function.rs:259
  56:     0x7ff7034c3736 - std::panicking::try::do_call
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:403
  57:     0x7ff7034c3736 - std::panicking::try
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:367
  58:     0x7ff7034c3736 - std::panic::catch_unwind
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panic.rs:129
  59:     0x7ff7034c3736 - std::rt::lang_start_internal::closure$2
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\rt.rs:45
  60:     0x7ff7034c3736 - std::panicking::try::do_call
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:403
  61:     0x7ff7034c3736 - std::panicking::try
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:367
  62:     0x7ff7034c3736 - std::panic::catch_unwind
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panic.rs:129
  63:     0x7ff7034c3736 - std::rt::lang_start_internal
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\rt.rs:45
  64:     0x7ff70342115f - std::rt::lang_start<tuple$<> >
                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\std\src\rt.rs:62
  65:     0x7ff703422f46 - main
  66:     0x7ff7034d409c - invoke_main
                               at d:\a01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  67:     0x7ff7034d409c - __scrt_common_main_seh
                               at d:\a01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  68:     0x7fff2a3f7034 - BaseThreadInitThunk
  69:     0x7fff2a9a2651 - RtlUserThreadStart
thread panicked while panicking. aborting.
```

**Comment by pradt2** [comment]
Sorry, my bad. This happens only if the latest `wireguard.dll` library file is used (straight from the [wireguard-nt website](https://download.wireguard.com/wireguard-nt/)).

I confirm that it's all fine when I use the specific `wireguard.dll` file from the `examples/wireguard-nt/amd64`. So this is a non-issue.

Having said that, it's a bit worrying that already the new version of the `wireguard.dll` has broken backwards compatibility.

**Comment by pradt2** [comment]
I tried running `bindgen` on the `wireguard.h` distributed together with `wireguard.dll` (NOT the one you put together), however, I see that `bindgen` doesn't really produce anything useful as far as the methods are concerned. Have you also experienced this? Is this why you decided to rewrite wireguard.h by hand?

I'm willing to spend some time trying to make `bindgen` work with the original `wireguard.h` file, so that hopefully we can release new versions with little to no manual intervention (i.e. without needing to manually verify/rewrite `wireguard.h` before we feed it to `bindgen`). 

However, before I do this, I'd like to know how/why you decided to rewrite `wireguard.h` by hand, what you tried etc, so that I don't end up just duplicating the effort you've already put into this.

@TroyNeubauer

**Comment by TroyNeubauer** [comment]
Yes, I decided to manually create a wireguard.h to pass to bindgen because the official header file only has typedefs for the functions and not normal external definitions that bindgen likes. To create it I went through the documentation and wrote out the functions as a c-compiler would expect them. I messed around with bindgen a little bit but concluded that for the effort, the api is unlikely to change very much so I bailed and did it by hand. That would be awesome if you could come up with a better solution. One crazy idea I had is if getting declarations from the distributed header is hard, it might be worthwhile to scrape this page: https://git.zx2c4.com/wireguard-nt/about/. It has a decently standardized format.

Also it looks like `WireguardDeleteAdapter` was renamed to `WireGuardCloseAdapter` in 0.10, which is why `GetProcAddress` is failing. I'll build out some logic for calling the right version of the function based on which one is available.
I'd be happy to accept any PR that makes keeping things up to date easy.

**Comment by pradt2** [comment]
Seems like we can get it done quite easily if `bindgen` supports static variables with `--dynamic-loading`. It should be possible since `libloading`, which is what is used under the hood, already supports it. 

I've opened an issue about it [--dynamic-loading ignores static variables](https://github.com/rust-lang/rust-bindgen/issues/2113). Let's see if it gets any traction with the devs over there, if not I will look into it myself.

**Comment by pradt2** [comment]
Hi, I managed to get `bindgen` to work nicely with the original `wireguard.h` header file (without any changes to `bindgen` itself). This means that we can take `wireguard.h` and, without modifying it in any way, get workable Rust bindings.

I did try generating new bindings just to see how they would fit in with the existing `adapter.rs`, To nobody's surprise, there are a couple of issues like changed function names etc. Not too many though.

I would like to donate this work to the project. However, since it technically breaks compilation (too lazy to fix the `adapter.rs` myself lol), I can't open a PR to your `main` branch.

Maybe if you could create a separate branch for my changes (say, `driver-0.10.1`), I could open a PR and push my work to that branch, and then you could take it the rest of the way?

@TroyNeubauer

**Comment by TroyNeubauer** [comment]
This sounds great. `driver-0.10.1` is now ready for you on here

**Comment by pradt2** [comment]
Hey, thanks a bunch. I've just pushed my changes to the branch. For a brief description of that is (and what isn't) included, see the description of https://github.com/nulldotblack/wireguard-nt/pull/4

**Comment by pradt2** [comment]
Solved by  #5. Closing
